File: leetcode/202.happy-number/ts/happynumber.ts
const isHappy = (n: number): boolean => {
    const seenNumbers = new Set<number>()

    while (true) {
        n = sumSquares(n)
        if (n === 1) return true
        if (seenNumbers.has(n)) return false
        seenNumbers.add(n)
    }

};

const sumSquares = (n: number): number => {
    let total = 0

    while (n > 0) {
        const r = n % 10
        n = Math.floor(n / 10)
        total += r * r
    }

    return total
}

export default isHappy

File: leetcode/202.happy-number/go/happynumber.go
package happynumber

func isHappy(n int) bool {
	s := make(map[int]struct{})

	for {
		n = sumSquares(n)

		if n == 1 {
			return true
		}

		if _, exists := s[n]; exists {
			return false
		}

		s[n] = struct{}{}
	}
}

func sumSquares(n int) int {
	total := 0

	for n > 0 {
		r := n % 10
		total += r * r
		n /= 10
	}

	return total
}
